File: controllers/incomeCategories.js
const { IncomeCategory, Household, HouseholdMember, sequelize } = require('../models');

class IncomeCategoriesController {
  /**
   * Admin: GET all income categories for a given householdId (query param)
   * Ordered by sortOrder, includes Household name
   */
  async adminList(ctx) {
    try {
      const { householdId } = ctx.query;

      if (!householdId) {
        ctx.status = 400;
        ctx.body = { error: 'householdId query parameter is required' };
        return;
      }

      const categories = await IncomeCategory.findAll({
        where: { householdId },
        include: [
          {
            model: Household,
            attributes: ['id', 'name']
          }
        ],
        order: [['sortOrder', 'ASC']]
      });

      ctx.body = {
        success: true,
        categories
      };
    } catch (error) {
      console.error('Admin list income categories error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to retrieve income categories' };
    }
  }

  /**
   * Admin: POST create a new income category
   * Body: { name, icon, color, sortOrder, householdId }
   */
  async adminCreate(ctx) {
    try {
      const { name, nameHe, icon, color, sortOrder, householdId } = ctx.request.body;

      if (!name || !icon || !householdId) {
        ctx.status = 400;
        ctx.body = { error: 'name, icon, and householdId are required' };
        return;
      }

      const household = await Household.findByPk(householdId);
      if (!household) {
        ctx.status = 404;
        ctx.body = { error: 'Household not found' };
        return;
      }

      const category = await IncomeCategory.create({
        name,
        nameHe: nameHe || null,
        icon,
        color,
        sortOrder,
        householdId
      });

      ctx.status = 201;
      ctx.body = {
        success: true,
        category
      };
    } catch (error) {
      console.error('Admin create income category error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to create income category' };
    }
  }

  /**
   * Admin: PUT /:id update an income category
   * Body: { name, icon, color, sortOrder }
   */
  async adminUpdate(ctx) {
    try {
      const { id } = ctx.params;
      const { name, nameHe, icon, color, sortOrder } = ctx.request.body;

      const category = await IncomeCategory.findByPk(id);
      if (!category) {
        ctx.status = 404;
        ctx.body = { error: 'Income category not found' };
        return;
      }

      await category.update({ name, nameHe: nameHe || null, icon, color, sortOrder });

      ctx.body = {
        success: true,
        category
      };
    } catch (error) {
      console.error('Admin update income category error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to update income category' };
    }
  }

  /**
   * Admin: DELETE /:id delete an income category
   */
  async adminDelete(ctx) {
    try {
      const { id } = ctx.params;

      const category = await IncomeCategory.findByPk(id);
      if (!category) {
        ctx.status = 404;
        ctx.body = { error: 'Income category not found' };
        return;
      }

      await category.destroy();

      ctx.body = { success: true };
    } catch (error) {
      console.error('Admin delete income category error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to delete income category' };
    }
  }

  /**
   * Admin: PUT /reorder bulk update sortOrder
   * Body: { items: [{id, sortOrder}] }
   */
  async adminReorder(ctx) {
    try {
      const { items } = ctx.request.body;

      if (!items || !Array.isArray(items)) {
        ctx.status = 400;
        ctx.body = { error: 'items array is required' };
        return;
      }

      await sequelize.transaction(async (t) => {
        for (const item of items) {
          await IncomeCategory.update(
            { sortOrder: item.sortOrder },
            { where: { id: item.id }, transaction: t }
          );
        }
      });

      ctx.body = { success: true };
    } catch (error) {
      console.error('Admin reorder income categories error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to reorder income categories' };
    }
  }

  /**
   * App: POST create an income category from within the app
   * Body: { name, icon, color, householdId }
   */
  async appCreate(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { name, nameHe, icon, color, householdId } = ctx.request.body;

      if (!name || !icon || !householdId) {
        ctx.status = 400;
        ctx.body = { error: 'name, icon, and householdId are required' };
        return;
      }

      const membership = await HouseholdMember.findOne({
        where: { householdId, appUserId: appUser.id }
      });
      if (!membership) {
        ctx.status = 403;
        ctx.body = { error: 'Not a member of this household' };
        return;
      }

      const maxOrder = await IncomeCategory.max('sortOrder', { where: { householdId } });
      const sortOrder = (maxOrder ?? 0) + 1;

      const category = await IncomeCategory.create({
        name,
        nameHe: nameHe || null,
        icon: icon || 'label',
        color: color || '#4CAF50',
        sortOrder,
        householdId
      });

      ctx.status = 201;
      ctx.body = { success: true, category };
    } catch (error) {
      console.error('App create income category error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to create income category' };
    }
  }

  /**
   * App: GET income categories for a household
   * Query: householdId
   * Validates that the current app user is a member of the household
   */
  async appList(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { householdId } = ctx.query;

      if (!householdId) {
        ctx.status = 400;
        ctx.body = { error: 'householdId query parameter is required' };
        return;
      }

      // Validate that the app user is a member of the requested household
      const membership = await HouseholdMember.findOne({
        where: { householdId, appUserId: appUser.id }
      });

      if (!membership) {
        ctx.status = 403;
        ctx.body = { error: 'You are not a member of this household' };
        return;
      }

      const categories = await IncomeCategory.findAll({
        where: { householdId },
        order: [['sortOrder', 'ASC']]
      });

      ctx.body = {
        success: true,
        categories
      };
    } catch (error) {
      console.error('App list income categories error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to retrieve income categories' };
    }
  }
}

module.exports = new IncomeCategoriesController();

File: backend_go/internal/handlers/income_categories.go
package handlers

import (
	"household-go/internal/database"
	"household-go/internal/models"
	"household-go/internal/utils"

	"github.com/gin-gonic/gin"
)

// IncomeCategoriesHandler mirrors ExpenseCategoriesHandler for income categories.
type IncomeCategoriesHandler struct{}

func NewIncomeCategoriesHandler() *IncomeCategoriesHandler {
	return &IncomeCategoriesHandler{}
}

func (h *IncomeCategoriesHandler) AdminList(c *gin.Context) {
	var categories []models.IncomeCategory
	database.DB.Order("sort_order ASC").Find(&categories)
	utils.OK(c, gin.H{"success": true, "categories": categories})
}

func (h *IncomeCategoriesHandler) AdminCreate(c *gin.Context) {
	var body struct {
		Name        string  `json:"name"        binding:"required"`
		NameHe      *string `json:"nameHe"`
		Icon        *string `json:"icon"`
		Color       *string `json:"color"`
		SortOrder   int     `json:"sortOrder"`
		HouseholdID uint    `json:"householdId" binding:"required"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "Name and householdId are required")
		return
	}
	cat := models.IncomeCategory{
		Name:        body.Name,
		NameHe:      body.NameHe,
		Icon:        body.Icon,
		Color:       body.Color,
		SortOrder:   body.SortOrder,
		HouseholdID: body.HouseholdID,
	}
	database.DB.Create(&cat)
	utils.Created(c, gin.H{"success": true, "category": cat})
}

func (h *IncomeCategoriesHandler) AdminUpdate(c *gin.Context) {
	id := c.Param("id")
	var cat models.IncomeCategory
	if err := database.DB.First(&cat, id).Error; err != nil {
		utils.NotFound(c, "Category not found")
		return
	}
	var body struct {
		Name      *string `json:"name"`
		NameHe    *string `json:"nameHe"`
		Icon      *string `json:"icon"`
		Color     *string `json:"color"`
		SortOrder *int    `json:"sortOrder"`
	}
	c.ShouldBindJSON(&body)
	updates := map[string]interface{}{}
	if body.Name != nil      { updates["name"] = *body.Name }
	if body.NameHe != nil    { updates["name_he"] = body.NameHe }
	if body.Icon != nil      { updates["icon"] = body.Icon }
	if body.Color != nil     { updates["color"] = body.Color }
	if body.SortOrder != nil { updates["sort_order"] = *body.SortOrder }
	if len(updates) > 0 {
		database.DB.Model(&cat).Updates(updates)
	}
	utils.OK(c, gin.H{"success": true, "category": cat})
}

func (h *IncomeCategoriesHandler) AdminDelete(c *gin.Context) {
	id := c.Param("id")
	var cat models.IncomeCategory
	if err := database.DB.First(&cat, id).Error; err != nil {
		utils.NotFound(c, "Category not found")
		return
	}
	database.DB.Delete(&cat)
	utils.OK(c, gin.H{"success": true})
}

func (h *IncomeCategoriesHandler) AppList(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)
	householdID := c.Query("householdId")
	if householdID == "" {
		utils.BadRequest(c, "householdId is required")
		return
	}
	if err := verifyMembership(householdID, appUser.ID); err != nil {
		utils.Forbidden(c, "Not a member of this household")
		return
	}
	var categories []models.IncomeCategory
	database.DB.Where("household_id = ?", householdID).Order("sort_order ASC").Find(&categories)
	utils.OK(c, gin.H{"success": true, "categories": categories})
}

func (h *IncomeCategoriesHandler) AppCreate(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)
	var body struct {
		Name        string  `json:"name"        binding:"required"`
		NameHe      *string `json:"nameHe"`
		Icon        *string `json:"icon"`
		Color       *string `json:"color"`
		HouseholdID uint    `json:"householdId" binding:"required"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "Name and householdId are required")
		return
	}
	if err := verifyMembership(formatUint(body.HouseholdID), appUser.ID); err != nil {
		utils.Forbidden(c, "Not a member of this household")
		return
	}
	cat := models.IncomeCategory{
		Name:        body.Name,
		NameHe:      body.NameHe,
		Icon:        body.Icon,
		Color:       body.Color,
		HouseholdID: body.HouseholdID,
	}
	database.DB.Create(&cat)
	utils.Created(c, gin.H{"success": true, "category": cat})
}
